### ISSUE TITLE
Mixing Datasets

### ISSUE THREAD

**Issue by Srini-98** [mosaicml/streaming#342]
Hi,

I was wondering if its possible to create a dataset by taking examples from different sub datasets and mixing them together using the streaming dataset. 

A bit like:

https://huggingface.co/docs/datasets/process#interleave

I tried using 'proportion' parameter but this still iterates one sub dataset at a time.

**Comment by knighton** [comment]
You need to set `shuffle=True`, it defaults to `False`.

**Comment by karan6181** [comment]
@Srini-98 Does the above solution worked for you?

**Comment by Srini-98** [comment]
@karan6181 @knighton I set shuffle to be true but when I am iterating it still goes only one sub dataset at a time. For context I am sharing  the sample code:

proportion_lis = [ ('code' , 0.50) , ('language' , 0.50)] 
streams = [Stream(local=path+"train_" + i[0] , proportion=i[1]) for i in proportion_lis]

dataset = StreamingDataset(streams=streams , shuffle=True)

for i in range(0 , 300):
    sample = dataset[i]
    print(sample)

When i switch the items in the proportion_lis list and run then I get language to come through but code doesnt. But what I want is the same batch to have both code and language. Any tips regarding this?

**Comment by Srini-98** [comment]
@karan6181 @knighton Yes this solution worked for me, thanks a lot. I have a doubt regarding resuming training with mixing datasets. Currently, I am planning to mix the datasets on the fly with the Streamingdataset (I have converted all my datasets to mds format). If there is a node failure or training breaks down for some reason, will I be able to resume the training instantaneously. I see two methods of implementing this:

1. Mix the datasets by shuffling and write to mds format before training starts. 
2. Mix on the fly during training by using proportion and shuffling. 

Will I be able to resume training instantaneously after a crash using method 2? Would there be an overhead due to the shuffling that takes place on the fly which the dataloader has to iterate through to reach the last point of training.

**Comment by karan6181** [comment]
Hi @Srini-98, Yes, you would be able to resume the training if you have enabled the checkpoint in your code. Streaming Dataset store the meta-data information as part of checkpointing. If you are using [composer](https://github.com/mosaicml/composer/), you don't need to do any extra step regarding resuming from the checkpoint, composer handles it under the hood. However, if you are not using composer, then you need to use [StreamingDataLoader](https://github.com/mosaicml/streaming/blob/main/streaming/base/dataloader.py#L16) which has a method `state_dict()` and a `load_state_dict()` which saves/load the meta-data information about the dataset which you would need to call manually during checkpointing. One of the example can be found [here](https://github.com/mosaicml/streaming/blob/main/tests/test_streaming.py#L297).

**Comment by karan6181** [comment]
@Srini-98 Closing this issue for now assuming it is being resolved. Please feel free to re-open it otherwise. Thank you!
